xwinglib updates for loading space objects

Dark-Visor opened the pull request.

Dark-Visor committed:
lib files should be close missionparse is not close

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -74,6 +74,23 @@ struct xwi_flightgroup {
 };
 #pragma pack(pop)
 
+struct xwi_objectgroup {
```
Move the `#pragma pack(pop)` line below the new struct.  The pack(push) and pack(pop) commands should enclose the structs.

Goober5000 commented on code/mission/import/xwinglib.h:
```diff
@@ -138,6 +138,51 @@ enum class XWMObjective : short
+{
+	ojf_Flat,
+	ojf_Edge,
+	ojf_Broadside,
```
Where did you get these values?  From the file format reference I see
```
00	Floor (X-Y plane)
01	Side (Y-Z plane)
02	Front (X-Z plane)
```

Dark-Visor replied:
ah I was taking those from the older (archived) reference. I'll update

Goober5000 commented on code/mission/import/xwinglib.h:
```diff
@@ -138,6 +138,51 @@ enum class XWMObjective : short
+	ojf_Edge,
+	ojf_Broadside,
+	ojf_Scattered,
+	ojf_Destroy
```
As I said in the DM:
> The mission goal will need to be split off and stored in a separate field in the friendly struct

So `ojf_Destroy` should not be included here

Goober5000 commented on code/mission/import/xwinglib.h:
```diff
@@ -210,6 +255,22 @@ class XWMObject
+	std::string cargo;
+	std::string specialCargo;
+
+	int specialObjectNumber;
```
Since these fields aren't actually used, then you might as well take them out.

Goober5000 commented on code/mission/import/xwinglib.h:
```diff
@@ -210,6 +255,22 @@ class XWMObject
+
+	XWMObjectType objectType;
+
+	XWMCraftIFF objectIFF; // shares with FlightGroup
```
Since mines are always hostile, and IFF is not relevant to other objects, you might as well take the IFF out too.

Goober5000 commented on code/mission/import/xwinglib.h:
```diff
@@ -210,6 +255,22 @@ class XWMObject
+
+	XWMCraftIFF objectIFF; // shares with FlightGroup
+
+	XWMObjectFormation formation;
```
You'll need to add a separate enum (survive/destroyed/neither) to handle the goal.

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
 
+	for (int n = 0; n < h->number_of_objects; n++) {
+		xwi_objectgroup *oj =
+			(xwi_objectgroup*)(data + sizeof(xwi_header) + sizeof(xwi_flightgroup) + sizeof(xwi_objectgroup) * n);
```
The `sizeof(xwi_flightgroup)` should be multiplied by `h->number_of_flight_groups` because the object data starts after *all* the flight groups have been parsed.

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
+		noj->specialCargo = oj->special_cargo;
+		noj->specialObjectNumber = oj->special_object_number;
+
+		switch (oj->object_type) {
```
Look at the `CraftType` list in the file format.  Mine1 is hexadecimal 12 (decimal 18) and the other object types likewise need their indexes fixed.

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
+				break;
+		}
+
+		switch (oj->object_formation) {
```
See the other review comments about object formation.  Note the special handling of bits 3 and 4: these are handled *independently*, so they need special bitwise checks.

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
+		noj->object_z = oj->object_z / 160.0f;
+		noj->object_yaw = oj->object_yaw / 160.0f;
+		noj->object_pitch = oj->object_pitch / 160.0f;
+		noj->object_roll = oj->object_roll / 160.0f;
```
The yaw/pitch/roll should not be divided by 160.  The pitch, though, should subtract 90.  Again, consult the file format reference.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,144 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
 	}
 }
 
+void parse_xwi_objectgroup(mission *pm, const XWingMission *xwim, const XWMObject *oj)
```
Split off the xwingmissionparse changes into a separate branch; they should not be part of this PR.  We'll cross that bridge when we come to it.

Dark-Visor committed:
lib file updates and corrections for Objects

Change missionparse.cpp so this is purely updates to the
xwinglib.cpp and xwinglib.h files in
relation to the import of objects.

Remove redundant entries and values that do not apply to objects. Update
naming conventions in line with more recent XWI docs. Separate goals
from formation.

Goober5000 commented on code/mission/import/xwinglib.h:
```diff
@@ -138,6 +138,67 @@ enum class XWMObjective : short
+{
+	ojg_Survive,
+	ojg_Destroyed,
+	ojg_Neither
```
Put `ojg_Neither` as the first item so it is assigned the value 0

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -72,6 +72,20 @@ struct xwi_flightgroup {
 	short secondary_target;
 };
+
+struct xwi_objectgroup {
```
The `xwi_objectgroup` struct must exactly match the file format reference, even if the `XWMObject` struct has more fields or fewer fields.  That's why I called `XWMObject` the "friendly" struct, because it is friendlier to what the user might do with it.  By contrast, you can't add or remove fields to/from `xwi_objectgroup` because the bytes will be misaligned.

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -582,5 +596,186 @@ bool XWingMission::load(XWingMission *m, const char *data)
+				return false;
+		}
+
+		switch (oj->object_formation) {
```
The formation code here will need to mask off the non-relevant bytes.

Goober5000 commented on code/mission/import/xwinglib.cpp:
```diff
@@ -582,5 +596,186 @@ bool XWingMission::load(XWingMission *m, const char *data)
+				return false;
+		}
+
+		switch (oj->object_goal) {
```
There is no `object_goal` field in the byte structure (see the comment about `xwi_objectgroup`).  Instead `noj->objectGoal` must be set using the bit checks.

Dark-Visor committed:
Update do BitWise operations to lib files

Update the xwinglib.cpp so that struct xwi_objectgroup would contain all
the necessary entries. Also update the formation and goal BitWise reads
so that they are read correctly.

Goober5000 reviewed (changes requested):
Excellent work.  This is nearly ready.

There's one more thing we should do.  If the object is any type of training platform, we should skip over the regular formation handling and just assign a formation of FloorXY and a goal of Neither.  According to the file format reference, the `object_formation` field controls which guns are present on the training platform, and how many seconds are on the clock for training missions.  This would be annoying and tedious to code, for not much gain.  Put a TODO there, and we can come back to it in the future if anyone desperately wants to import training missions.

Dark-Visor committed:
Update xwinglib.cpp in prep for trainingplatforms

Object formation is split between training platforms and other objects.
This avoids redundant processing of formations for traning platforms
finding formations which they shouldn't have. A comment TODO also
prepares for later evaluting the object_formation to determine gun
placements and timing countdown for training platforms.

Goober5000 reviewed (approved):
Fantastic work 👍
